src/envoy/tcp/sni_verifier/sni_verifier.cc

```diff
@@ -0,0 +1,161 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/envoy/tcp/sni_verifier/sni_verifier.h"
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/exception.h"
+#include "envoy/network/connection.h"
+#include "envoy/stats/scope.h"
+
+#include "common/common/assert.h"
+
+#include "openssl/bytestring.h"
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Tcp {
+namespace SniVerifier {
+
+Config::Config(Stats::Scope& scope, uint32_t max_client_hello_size)
+    : stats_{SNI_VERIFIER_STATS(POOL_COUNTER_PREFIX(scope, "sni_verifier."))},
+      ssl_ctx_(SSL_CTX_new(TLS_with_buffers_method())),
+      max_client_hello_size_(max_client_hello_size) {
+  if (max_client_hello_size_ > TLS_MAX_CLIENT_HELLO) {
+    throw EnvoyException(fmt::format(
+        "max_client_hello_size of {} is greater than maximum of {}.",
+        max_client_hello_size_, size_t(TLS_MAX_CLIENT_HELLO)));
+  }
+
+  SSL_CTX_set_options(ssl_ctx_.get(), SSL_OP_NO_TICKET);
+  SSL_CTX_set_session_cache_mode(ssl_ctx_.get(), SSL_SESS_CACHE_OFF);
+  SSL_CTX_set_tlsext_servername_callback(
+      ssl_ctx_.get(), [](SSL* ssl, int* out_alert, void*) -> int {
+        SniVerifierFilter* filter =
+            static_cast<SniVerifierFilter*>(SSL_get_app_data(ssl));
+        filter->onServername(
+            SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name));
+
+        // Return an error to stop the handshake; we have what we wanted
+        // already.
+        *out_alert = SSL_AD_USER_CANCELLED;
+        return SSL_TLSEXT_ERR_ALERT_FATAL;
+      });
+}
+
+bssl::UniquePtr<SSL> Config::newSsl() {
+  return bssl::UniquePtr<SSL>{SSL_new(ssl_ctx_.get())};
+}
+
+thread_local uint8_t SniVerifierFilter::buf_[Config::TLS_MAX_CLIENT_HELLO];
+
+SniVerifierFilter::SniVerifierFilter(const ConfigSharedPtr config)
+    : config_(config), ssl_(config_->newSsl()) {
+  RELEASE_ASSERT(sizeof(buf_) >= config_->maxClientHelloSize(), "");
+
+  SSL_set_app_data(ssl_.get(), this);
```

Comment:
Not sure when and how the callback is called.  If it maybe called outside of SSL_do_handshake() call, then it is dangerous, the filter may go away when it is called.   

If it is always called inside SSL_do_handshake(), then it is better to set the app_data and call_back right before the SSL_do_handshake() and reset it after.